core/src/main/scala/kafka/controller/TopicDeletionManager.scala

```diff
@@ -102,8 +102,12 @@ class TopicDeletionManager(controller: KafkaController,
    */
   def enqueueTopicsForDeletion(topics: Set[String]) {
     if (isDeleteTopicEnabled) {
-      topicsToBeDeleted ++= topics
       partitionsToBeDeleted ++= topics.flatMap(controllerContext.partitionsForTopic)
+      // move the respective partitions to OfflinePartition and NonExistentPartition state, so that the
+      // offlinePartitionCount is properly updated
+      controller.partitionStateMachine.handleStateChanges(partitionsToBeDeleted.toSeq, OfflinePartition)
+      controller.partitionStateMachine.handleStateChanges(partitionsToBeDeleted.toSeq, NonExistentPartition)
```

Comment:
Hmm, moving the code here fixes the offlinePartitionCount issue. However, it introduces another issue. If the topic is in pending deletion state because of an going partition reassignment, the partition can't be moved to online state once it's in NonExistentPartition state. This may prevent the reassignment from completing.

Reply:
Good catch!
To address this issue, the latest commit in this PR has the following changes
- Removed the TopicDeletionManager#partitionsToBeDeleted variable since it is always updated together with the topicsToBeDeleted variable, and therefore checking whether a partition is to be deleted can be done through the topicsToBeDelete variable.
- Added a new variable partitionsBeingDeleted, which includes partitions that are going through actual deletions. More partitions are added to it in the TopicDeletionManager#onTopicDeletion method, where the deletion for a partition has started. Once a partition is added to the partitionsBeingDeleted variable, we are sure partition reassignment can NO LONGER happen to it.
- The PartitionStateMachine#updateControllerMetrics is updated to check if a partition is BEING deleted. This means if reassignment is going on for a given partition while the corresponding topic is enqueued for deletion, we will still monitor its state change until the reassignment finishes and the deletion starts.
Let me know what you think, and thanks a lot for your review!